### PR TITLE
Add return_all_channels option to perform_forecasting() (fixes #25)

### LINKED ISSUE #25: [Feature Request] Add multivariate (multi-target) output support to perform_forecasting() — backbone already produces [B, C, H], SDK truncates to single target
## What's happening

`perform_forecasting()` accepts multivariate input but returns only a single `{target_column}_forecast`. The backbone already produces `[B, C, H]` — the SDK just drops `C-1` channels at `sdk/forecasting.py` L1915:

```python
forecast_values = P_orig[:forecast_horizon, 0]   # ← single channel extraction
```

`P_orig_reshaped` is computed with shape `[B, C, H]` (L1897) and then thrown away. The Cross-Channel head (`run8_best_model_cr.pt`) is designed to learn inter-channel correlations and outputs all channels in one forward pass.

## Why it matters

On a 10-channel forecasting use case (V=10, context=512, horizon=16, n=50, fine-tuned + cross-channel checkpoint, DGX Spark GB10 / CUDA 13):

| Path                                    | p95 latency |
| --------------------------------------- | ----------- |
| `perform_forecasting()` × 10 loop       | 402 ms      |
| `autoregressive_forecast()` direct call | 34 ms       |

About 12× speed-up just from stopping the V-loop. The model already does the work.

## Proposal

Add a `return_all_channels: bool = False` flag (or similar name) to `perform_forecasting()`:

- `False` (default): unchanged
- `True`: return one `{col}_forecast` column per feature

Reuses the existing `P_orig_reshaped` tensor; no new model calls.

## Open questions

1. Parameter name preference (`return_all_channels` / `return_multivariate` / `multi_target` / other)?
2. Should the API also accept an explicit `target_columns: list[str]` subset?
3. DARR mode kNN blending — uniform alpha/k per channel, or per-channel?
4. Standardizer file format — record channel order explicitly to avoid silent mismatches?

Happy to send a draft PR. Thanks for open-sourcing this.

### PR DESCRIPTION
**Pull request by himorishige**
Fixes #25.

Adds `return_all_channels: bool = False` to `perform_forecasting()` so the SDK can return the full `[B, C, H]` forecast that the backbone already produces, instead of discarding `C-1` channels at L1915.

## Change

- Default (`False`): unchanged. 38 existing tests pass.
- `True`: returns one `{col}_forecast` per feature column, in `columns_to_process` order.

Reuses the already-computed `P_orig_reshaped` tensor — no new model calls.

## Benchmark

V=10, context=512, horizon=16, n=50, fine-tuned + cross-channel, DGX Spark (GB10, CUDA 13, torch 2.12.1+cu130):

| Path                                                  | p50    | p95    | 200 ms SLA |
| ----------------------------------------------------- | ------ | ------ | ---------- |
| Looping `perform_forecasting()` over 10 targets       | 392 ms | 402 ms | ✗          |
| Single call with `return_all_channels=True` (this PR) | ~34 ms | ~34 ms | ✓          |

Same accuracy (RMSE 1.06 vs 1.07).

## Testing

- `make lint` / `make spdx-check` — pass
- `cd forecasting && uv run pytest sdk/tests` — **38 passed**
- Smoke: `return_all_channels=True` → shape `(16, 11)` with all 10 `{col}_forecast` columns; default → shape `(16, 2)` (backward-compatible)

## Notes

- Naming: `return_all_channels` is my pick but happy to rename — see Issue #1 for alternatives.
- DARR and Interpretability modes are not touched in this PR; both share the same `preds → P_orig` pipeline so they continue to work. Multivariate output for those can be a follow-up.
- Open design questions are listed in Issue #1 — happy to address any of them here or separately.

### CODE REVIEW

**Review by agautam478 — CHANGES_REQUESTED** [review]
Thanks for adding this. The feature direction makes sense, but I think this needs a few fixes before merge.

1. `return_all_channels` is inserted in the middle of the public `perform_forecasting()` signature, before `standardizer_pkl`. That is backward-incompatible for positional callers. Please move it to the end of the signature, ideally after `softmax_tau`, so existing positional usage keeps working.

2. `return_all_channels=True` is silently ignored when `interpretability=True`, because that path returns before the new output branch runs. Please either support all-channel interpretability output or raise a clear `ValueError` when both flags are used.

3. The new public option needs docs and test coverage. `forecasting/sdk/README.md` still documents only target-column output, and there is no `return_all_channels` test. Please add a standard-mode test that asserts `target_forecast` plus feature forecast columns, and update the SDK README signature/output section.

Local validation on a synthetic merge with current `main`: `make lint` passed, `pytest sdk/tests/test_forecasting.py -q` passed with 41 tests, and a mocked smoke test for `return_all_channels=True` returned `timestamp`, `target_forecast`, and `feature_forecast` as expected.

**Comment by agautam478** [comment]
One additional note on channel mixing: this feature is conceptually aligned with the cross-channel checkpoint because the model already predicts `[B, C, H]`, but it makes channel identity/order more important.

For `return_all_channels=True`, please document that inference columns must match the checkpoint/standardizer training order. With channel mixing, if the same columns are provided in a different order, or if extra/missing numeric columns change the channel layout, the model can still run but the forecast columns may be semantically wrong.

This does not need to block the minimal API shape, but we should treat explicit channel-order metadata/validation in the standardizer artifact as a follow-up improvement.

**Comment by himorishige** [comment]
Thanks for the detailed review! Addressed all three points and rebased onto current `main`:

1. **Signature position**: Moved `return_all_channels` to the end of the signature (after `softmax_tau`), so existing positional callers are unaffected.
2. **`interpretability` combo**: Combining `return_all_channels=True` with `interpretability=True` now raises a `ValueError` up front instead of silently ignoring the flag. I went with the explicit error rather than all-channel interpretability output, since the attribution pipeline is target-aligned — happy to explore per-channel interpretability separately if that would be useful.
3. **Docs + tests**: Updated `sdk/README.md` (function signature, key parameters, and outputs sections) and added five tests: per-channel value assertions with a channel-indexed stub model, the autoregressive path (`forecast_horizon > model_horizon`), the DARR-mode combination, and the two `ValueError` guards.

While revisiting this I also tightened the output path itself:

- Unified the single-channel and all-channel result assembly into one loop, which fixes a dtype inconsistency (the all-channel path returned float32 columns while the default path returns float64 via `.tolist()`).
- Added a guard for `timestamp_column` colliding with an emitted `{column}_forecast` name (previously a silent dict-key overwrite).
- The completion log now lists the actual emitted columns instead of always claiming only `{target_column}_forecast`.

Validation on the rebased branch: `make lint`, `make spdx-check`, and `pytest sdk/tests/test_forecasting.py -q` (46 passed).

One heads-up found while testing DARR + `return_all_channels`: the existing column-alignment check compares column *sets*, so an input/context pair holding the same numeric columns in a different order blends misaligned channels. That is pre-existing and independent of this PR, but the new flag widens its blast radius since all channels become visible — I'll file a separate issue with a repro.

**Review by agautam478 — APPROVED** [review]
Thanks for the updates. The previous blockers are addressed: `return_all_channels` is now at the end of the public signature, the unsupported `interpretability=True` combination raises clearly, and the docs/tests cover the new behavior including DARR and autoregressive paths.

Local validation on the updated branch: `make lint` passed, `make spdx-check` passed, and `pytest sdk/tests/test_forecasting.py -q` passed with 46 tests.

Non-blocking nit before/while merging: the PR body and first commit still mention `Issue #1`; those should say `#25` or the new follow-up `#50` to avoid confusing future readers.

I agree with keeping the DARR channel-order fix in #50 as a separate follow-up.
